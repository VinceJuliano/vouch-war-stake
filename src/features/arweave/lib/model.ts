import { z } from "zod";

export const ArweaveId = z.string().regex(/^[a-zA-Z0-9_-]{43}$/);
export type ArweaveId = z.infer<typeof ArweaveId>;

export const ArweaveAddress = ArweaveId;
export type ArweaveAddress = ArweaveId;

export const ArweaveTxId = ArweaveId;
export type ArweaveTxId = ArweaveId;
