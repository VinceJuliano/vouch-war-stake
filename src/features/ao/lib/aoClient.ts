import { connect, result } from "@permaweb/aoconnect";

export type MessageId = string;

type BaseAoClient = ReturnType<typeof connect>;

export type AoClient = BaseAoClient;

export type MessageResult = Awaited<ReturnType<typeof result>>;

// export type Message = MessageResult["Messages"][0];
export type Message = {
  Target: string;
  Tags: Array<{
    name: string;
    value: string;
  }>;
  Data: string;
};

export type MessageWithResult = {
  messageId: MessageId;
  result: MessageResult;
};
