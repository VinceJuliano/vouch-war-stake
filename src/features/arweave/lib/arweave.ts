import { base32, base64urlnopad } from "@scure/base";
import Arweave from "arweave";

export const defaultArweave = Arweave.init({});

export function fetchUrl(txId: string) {
  const txIdData = base64urlnopad.decode(txId);
  const base32UnpaddedTxId = base32.encode(txIdData).replace(/=/g, "");
  return `https://${base32UnpaddedTxId}.arweave.net/${txId}`;
}
