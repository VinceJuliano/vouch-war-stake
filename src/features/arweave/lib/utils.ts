export const truncateAddress = (
  address: string,
  prefixLength: number = 6,
  suffixLength: number = 4,
  middleText: string = "...",
) => {
  return `${address.slice(0, prefixLength)}${middleText}${address.slice(-suffixLength)}`;
};
