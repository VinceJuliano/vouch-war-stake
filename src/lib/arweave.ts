export function getTagValue(tags: Tag[], name: string) {
  return tags.find((tag) => tag.name === name)?.value as string | undefined;
}
