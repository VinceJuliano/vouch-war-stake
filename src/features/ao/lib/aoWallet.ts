import { createDataItemSigner } from "@permaweb/aoconnect";
import { WalletType } from "./wallets";

export type AoSigner = ReturnType<typeof createDataItemSigner>;

export type AoWallet = {
  type: WalletType;
  anonymous: boolean;
  address: string;
  signer: AoSigner;
};

export type AoWalletConnectionResult =
  | {
      success: true;
      result: AoWallet;
      disconnect?: () => void;
    }
  | {
      success: false;
      error: unknown;
    };

// TODO: Import these from somewhere more official
export type PermissionType =
  | "ACCESS_ADDRESS"
  | "ACCESS_PUBLIC_KEY"
  | "ACCESS_ALL_ADDRESSES"
  | "SIGN_TRANSACTION"
  | "ENCRYPT"
  | "DECRYPT"
  | "SIGNATURE"
  | "ACCESS_ARWEAVE_CONFIG"
  | "DISPATCH";

export interface AppInfo {
  name?: string;
  logo?: string;
}

export interface GatewayConfig {
  host: string;
  port: number;
  protocol: "http" | "https";
}

export type ConnectConfig = {
  permissionsRequested: PermissionType[];
  permissionsRequired?: PermissionType[];
  appInfo?: AppInfo;
  gateway?: GatewayConfig;
};

export type AoWalletConnector = (
  config: ConnectConfig,
  onDisconnect?: () => void,
) => Promise<AoWalletConnectionResult>;
