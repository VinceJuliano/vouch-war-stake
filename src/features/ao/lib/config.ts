import { PermissionType } from "./aoWallet";

export const permissionsRequired: Array<PermissionType> = [
  "ACCESS_ADDRESS",
  "SIGN_TRANSACTION",
  // "ACCESS_PUBLIC_KEY",
  // "ENCRYPT",
  // "DECRYPT",
  "SIGNATURE",
  "ACCESS_ARWEAVE_CONFIG",
];

export const permissionsRequested = [
  ...permissionsRequired,
  // ...optionalPermissions,
];
