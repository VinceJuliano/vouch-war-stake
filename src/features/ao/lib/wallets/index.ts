import { AoWalletConnector } from "../aoWallet";
import { connectInjectedWallet } from "./injected";
import { connectOthentWallet } from "./othent";
import { createGeneratedWallet } from "./generated";

export const WalletType = ["Injected", "Othent", "Keyfile"] as const;
export type WalletType = (typeof WalletType)[number];

export const wallets: Record<WalletType, AoWalletConnector> = {
  Injected: connectInjectedWallet,
  Othent: connectOthentWallet,
  Keyfile: createGeneratedWallet,
} as const;
