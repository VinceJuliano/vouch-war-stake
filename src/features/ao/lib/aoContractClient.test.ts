import { describe, expect, test } from "vitest";
import { createAoContractClient } from "./aoContractClient";
import { loadTestWallet } from "../test/lib/fsWallet";
import { connect } from "@permaweb/aoconnect";

describe("createAoContractClient", () => {
  test("creates client", async () => {
    const wallet = await loadTestWallet();

    const aoClient = connect();
    const aoContractClient = createAoContractClient(
      import.meta.env.VITE_READ_PROCESS_ID,
      aoClient,
      wallet,
    );

    expect(aoContractClient).toMatchSnapshot();
  });
});
