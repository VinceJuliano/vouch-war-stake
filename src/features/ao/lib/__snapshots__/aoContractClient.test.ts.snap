// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createAoContractClient > creates client 1`] = `
{
  "aoClient": {
    "assign": [Function],
    "dryrun": [Function],
    "message": [Function],
    "monitor": [Function],
    "result": [Function],
    "results": [Function],
    "spawn": [Function],
    "unmonitor": [Function],
  },
  "aoWallet": {
    "address": "HHz5nNVEzkALjmZq8cdIdubMBJCpIFOvzO5uHfU5nQI",
    "anonymous": false,
    "signer": [Function],
    "type": "Keyfile",
  },
  "dryrunReadReplyOne": [Function],
  "dryrunReadReplyOneJson": [Function],
  "dryrunReadReplyOptional": [Function],
  "message": [Function],
  "processId": "PHHOfjFPSbn7TQT1Z-dB4ndyIyppHV5haMPCg8dIccs",
}
`;
